src/inmanta/agent/cache.py

```diff
@@ -63,51 +62,33 @@ def __init__(
             a callable expecting the cached value as an argument.
         """
         self.key = key
-        self.scope = scope
+        self.scope: Scope = scope
         self.value = value
-        self.time: float = time.time() + scope.timeout
         self.call_on_delete = call_on_delete
+        self.expiry_time = time.time() + scope.timeout
+        self.called_finalizer = False
 
     def __lt__(self, other: "CacheItem") -> bool:
-        return self.time < other.time
+        return self.expiry_time < other.expiry_time
 
     def delete(self) -> None:
-        if callable(self.call_on_delete):
+        if callable(self.call_on_delete) and not self.called_finalizer:
+            self.called_finalizer = True
             self.call_on_delete(self.value)
 
     def __del__(self) -> None:
         self.delete()
 
 
-class CacheVersionContext(contextlib.AbstractContextManager):
-    """
-    A context manager to ensure the cache version is properly closed
-    """
-
-    def __init__(self, cache: "AgentCache", version: int) -> None:
-        self.version = version
-        self.cache = cache
-
-    def __enter__(self):
-        self.cache.open_version(self.version)
-        return self.cache
-
-    def __exit__(self, exc_type, exc_val, exc_tb):
-        self.cache.close_version(self.version)
-        return None
-
-
 @stable_api
 class AgentCache:
     """
     Caching system for the agent:
 
-    cache items can expire based on:
-    1. time
-    2. version
+    cache items are removed from the cache either:
+    1. individually when their expiry_time is up
+    2. as a group when their version is not being used for a while (default 1 min)
```

Comment:
1. Agree
2. 
> for which for_version=True was set  

I think `for_version` is always implicitly set to True, maybe more accurate would be "for which version=x was set"?
> we want to expire phased out versions' cache entries even if their lifetime has not ended yet, in order to keep the memory footprint acceptable

I agree
> And "last used" refers to the version being used by the agent (cache.open_version() or similar), **rather than any of its cache entries being used**

Why not ? In my mind these are equivalent (any of its cache entries _for this version_ being used)

Reply:
> always implicitly set to True

Good catch, I forgot to check and just assumed the default was `False`. That revelation makes this whole thing more relevant, though it still all depends on whether the user passes the version or not.

We could consider to change the default, because I can imagine its primary purpose used to be to allow new code to recalculate the value, but our executors already achieve that now (not entirely sure of the in-process ones, but I think even there each executor has their own cache?).

Or we can even go as far as to drop the `for_version` argument entirely, which might be what was meant with the proposal in the ticket (on the other hand, then the version-based timer suggestion wouldn't make sense anymore).

> Why not ? In my mind these are equivalent (any of its cache entries for this version being used)

I would think we'd want to keep a version's cache alive as long as the version is active, even if none of the resources that were deployed in the past minute happened to read from the cache. e.g. imagine an agent with a few resources of type `X` and many of type `Y`. The ones of type `X` use the cache. They are deployed in order `x0 y00 y01 ... x1 y10 y11 ...`. Suppose the deploy of `y00 .. y0n` takes more than a minute, I would think we'd want to keep the cache for `X` alive, because we're still actively deploying the version, and `x1` will need it.

Reply:
I find it difficult to decide on the best approach. Its tough to trust to our judgement when it's clear we don't fully understand what exactly Wouter suggested and why. I'll get back to this.

Reply:
> I would think we'd want to keep the cache for X alive, because we're still actively deploying the version, and x1 will need it

I agree

Reply:
I discussed this with Wouter earlier today. Key takeaways:
- the general reasoning above is correct.
- keep the default for `for_version`
- for cache entries with `for_version=False`, use only the expiry time (time since creation)
- for cache entries with `for_version=True`, add a linger time: time since last use. No need to keep track of this per version, also no need to keep track of this since agent active. It suffices to track last access of the cache value.
- The cache should only expire while the agent is inactive. e.g. before each deploy (or other resource action) starts, freeze the cache, afterwards unfreeze it. While unfrozen, the periodic check is active. When frozen, the check is run once more and then paused until unfrozen. This ensures consistent cache entries for at least a single resource deploy. The deploy will then update any "last used" timers for cache entries it consumes.

@wouterdb is that the gist of it?

Reply:
sounds good to me